Implementation: `#[feature(sync_nonpoison)]`, `#[feature(nonpoison_mutex)]`

connortsui20 opened the pull request:
Continuation of https://github.com/rust-lang/rust/pull/134663

Tracking Issue: https://github.com/rust-lang/rust/issues/134645

This PR implements a new `sync/nonpoison` module, as well as the `nonpoison` variant of the `Mutex` lock.

There are 2 main changes here, the first is the new `nonpoison::mutex` module, and the second is the `mutex` integration tests.

For the `nonpoison::mutex` module, I did my best to align it with the current state of the `poison::mutex` module. This means that several unstable features (`mapped_lock_guards`, `lock_value_accessors`, and `mutex_data_ptr`) are also in the new `nonpoison::mutex` module, under their respective feature gates. Everything else in that file is under the correct feature gate (`#[unstable(feature = "nonpoison_mutex", issue = "134645")]`).

Everything in the `nonpoison::mutex` file is essentially identical in spirit, as we are simply removing the error case from the original `poison::mutex`.

The second big change is in the integration tests. I created a macro called that allows us to duplicate tests that are "generic" over the different mutex types, in that the poison mutex is always `unwrap`ped.

~~I think that there is an argument against doing this, as it can make the tests a bit harder to understand (and language server capabilities are weaker within macros), but I think the benefit of code deduplication here is worth it. Note that it is definitely possible to generalize this (with a few tweaks) to testing the other `nonpoison` locks when they eventually get implemented, but I'll leave that for a later discussion.~~

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16329031559/job/46126645082) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46126645082) [(plain)](https://github.com/rust-lang/rust/commit/deda5c5abcdd2ad2884161d30e7186142f5c55af/checks/46126645082/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 8.364
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6164 files
Diff in /checkout/library/std/src/sync/nonpoison/mutex.rs:130:
 #[must_not_suspend = "holding a MappedMutexGuard across suspend \
                       points can cause deadlocks, delays, \
                       and cause Futures to not implement `Send`"]
-// #[unstable(feature = "mapped_lock_guards", issue = "117108")] 
+// #[unstable(feature = "mapped_lock_guards", issue = "117108")]
 #[unstable(feature = "nonpoison_mutex", issue = "134645")]
 #[clippy::has_significant_drop]
 pub struct MappedMutexGuard<'a, T: ?Sized + 'a> {
Diff in /checkout/library/std/src/sync/nonpoison/mutex.rs:166:
     pub const fn new(t: T) -> Mutex<T> {
         Mutex { inner: sys::Mutex::new(), data: UnsafeCell::new(t) }
     }
-    
+
     // TODO(connor): Add `lock_value_accessors` feature methods.
 }
 
Diff in /checkout/library/std/src/sync/nonpoison/mutex.rs:222:
     ///
     /// If the mutex could not be acquired because it is already locked, then
     /// this call will return [`None`].
-    /// 
+    ///
     /// TODO(connor): This should return a `TryLockResult` as specified in
     /// <https://github.com/rust-lang/rust/issues/134645>
     ///
Diff in /checkout/library/std/src/sync/nonpoison/mutex.rs:289:
     pub fn get_mut(&mut self) -> &mut T {
         self.data.get_mut()
     }
-    
+
     // TODO(connor): Add `mutex_data_ptr` feature method.
 }
 
Diff in /checkout/library/std/src/sync/nonpoison/mutex.rs:381:
     /// This is an associated function that needs to be used as
     /// `MutexGuard::map(...)`. A method would interfere with methods of the
     /// same name on the contents of the `MutexGuard` used through `Deref`.
-    // #[unstable(feature = "mapped_lock_guards", issue = "117108")] 
+    // #[unstable(feature = "mapped_lock_guards", issue = "117108")]
     #[unstable(feature = "nonpoison_mutex", issue = "134645")]
     pub fn map<U, F>(orig: Self, f: F) -> MappedMutexGuard<'a, U>
     where
Diff in /checkout/library/std/src/sync/nonpoison/mutex.rs:407:
     /// `MutexGuard::try_map(...)`. A method would interfere with methods of the
     /// same name on the contents of the `MutexGuard` used through `Deref`.
     #[doc(alias = "filter_map")]
-    // #[unstable(feature = "mapped_lock_guards", issue = "117108")] 
+    // #[unstable(feature = "mapped_lock_guards", issue = "117108")]
     #[unstable(feature = "nonpoison_mutex", issue = "134645")]
     pub fn try_map<U, F>(orig: Self, f: F) -> Result<MappedMutexGuard<'a, U>, Self>
     where
Build completed unsuccessfully in 0:00:43
  local time: Wed Jul 16 20:01:44 UTC 2025
  network time: Wed, 16 Jul 2025 20:01:44 GMT
##[error]Process completed with exit code 1.
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16330008636/job/46130006280) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46130006280) [(plain)](https://github.com/rust-lang/rust/commit/19b1a2cf9582227d1c82a3f629777f203de9d286/checks/46130006280/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Checking tidy rustdoc_json...
Running eslint on rustdoc JS files
No error code explanation was removed!
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/std/src/sync/nonpoison/mutex.rs:180: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/library/std/src/sync/nonpoison/mutex.rs:236: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/library/std/src/sync/nonpoison/mutex.rs:303: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:19
  local time: Wed Jul 16 20:54:15 UTC 2025
  network time: Wed, 16 Jul 2025 20:54:15 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16331174423/job/46133920842) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46133920842) [(plain)](https://github.com/rust-lang/rust/commit/a25923227c4003acb6350b1796222f929d1e92b1/checks/46133920842/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Checking tidy rustdoc_json...
Running eslint on rustdoc JS files
No error code explanation was removed!
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/std/src/sync/nonpoison/mutex.rs:275: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/library/std/src/sync/nonpoison/mutex.rs:395: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:15
  local time: Wed Jul 16 22:02:50 UTC 2025
  network time: Wed, 16 Jul 2025 22:02:50 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

connortsui20 commented on library/std/tests/sync/mutex.rs:
The git diff here is somewhat unfortunate, as I had to reorder the tests for things to make sense. All of the "duplicated" tests here are taken from the existing tests in the file, so there is technically nothing new here other than the macro.

I would ideally like to write some more tests (testing more panic behavior in `nonpoison`, better tests for the `replace` drop behavior), but this is already quite a large PR.

tgross35 reviewed (changes requested):
Looks pretty good to me, most comments here are related to docs. This will need some history cleanup, there isn't any need to preserve the current state of https://github.com/rust-lang/rust/pull/134663 in history (I assume this was just done for now to show what changed since then)

connortsui20 commented on library/std/src/sync/nonpoison/mod.rs:
```diff
@@ -0,0 +1,37 @@
+    }
+}
+
+#[unstable(feature = "mapped_lock_guards", issue = "117108")]
+pub use self::mutex::MappedMutexGuard;
+#[unstable(feature = "nonpoison_mutex", issue = "134645")]
+pub use self::mutex::{Mutex, MutexGuard};
```
Yes this makes sense, though the only reason I did this is because [`poison.rs`](https://github.com/rust-lang/rust/blob/master/library/std/src/sync/poison.rs#L61-L75) does this as well and I couldn't find any comments explaining why. I can still remove this easily but should we also remove it in `poison.rs`?

tgross35 replied:
Actually it's probably fine to keep - I think this might be for sorting, so `rustfmt`'s `StdExternalCrate` grouping knows what's going on. Not that it really matters here.

connortsui20 commented on library/std/src/sync/nonpoison/mutex.rs:
```diff
@@ -0,0 +1,697 @@
+                d.field("data", &&*guard);
+            }
+            Err(WouldBlock) => {
+                d.field("data", &format_args!("<locked>"));
```
I'll change it in the original file as well

EFanZh commented on library/std/src/sync/nonpoison/mutex.rs:
```diff
@@ -0,0 +1,697 @@
+    #[unstable(feature = "lock_value_accessors", issue = "133407")]
+    pub fn set(&self, value: T) {
+        let old = self.replace(value);
+        if mem::needs_drop::<T>() {
+            drop(old)
+        }
```
I think `if` statement does not do anything. The original implementation uses `needs_drop` to decide whether to utilize `replace`, while this implementation will call `replace` unconditionally, potentially causing some `memcpy` overhead for trivially drop types.

connortsui20 committed:
clean up existing poison files

connortsui20 commented on library/std/src/lib.rs:
```diff
@@ -407,6 +407,7 @@
 #![feature(edition_panic)]
 #![feature(format_args_nl)]
 #![feature(log_syntax)]
+#![feature(nonpoison_mutex)]
```
Does anyone know if this supposed to be here? This is a remnant of the previous PR, and I don't really know what it is for. I've never seen any documentation on why these features need to be here either...

connortsui20 committed:
add `nonpoison::mutex` implementation

Adds the equivalent `nonpoison` types to the `poison::mutex` module.
These types and implementations are gated under the `nonpoison_mutex`
feature gate.

Also blesses the ui tests that now have a name conflicts (because these
types no longer have unique names). The full path distinguishes the
different types.

Co-authored-by: Aandreba <aandreba@gmail.com>
Co-authored-by: Trevor Gross <tmgross@umich.edu>

connortsui20 committed:
reorder mutex tests

This commit simply helps discern the actual changes needed to test both
poison and nonpoison locks.

connortsui20 committed:
add nonpoison and poison mutex tests

Adds tests for the `nonpoison::Mutex` variant by using a macro to
duplicate the existing `poison` tests.

Note that all of the tests here are adapted from the existing `poison`
tests.

connortsui20 commented on library/std/tests/sync/lib.rs:
```diff
@@ -29,3 +32,50 @@ mod rwlock;
+        #[test]
+        fn $name() {
+            #[allow(unused_imports)]
+            use ::std::sync::nonpoison::*;
```
I bring it in as `locks`

tgross35 commented on library/std/tests/sync/lib.rs:
```diff
@@ -29,3 +32,55 @@ mod rwlock;
+// no-op.
+//
+// The `poison` test will have the same `name`, but with a suffix of `_unwrap_poisoned`.
+#[macro_export]
```
You may need to add `pub(crate) use macro_name;` after the macro, which should let you use `super::macro_name` where needed. If that doesn't work, put the macro above the `mod` definition.

I don't fully understand macro scoping...

connortsui20 replied:
Ah it worked with just a `use nonpoison_and_poison_unwrap_test;`

And I don't understand macro scoping either haha

connortsui20 commented on library/std/tests/sync/lib.rs:
```diff
@@ -29,3 +32,55 @@ mod rwlock;
 #[path = "../common/mod.rs"]
 mod common;
+
+// A macro that generates two test cases for both the poison and nonpoison locks.
```
Actually I get an error when this is a doc comment, unless I should use `#[allow(unused_doc_comments)]`?

connortsui20 commented on library/std/src/sync/nonpoison.rs:
```diff
@@ -0,0 +1,37 @@
+
+/// A lock could not be acquired at this time because the operation would otherwise block.
+#[unstable(feature = "sync_nonpoison", issue = "134645")]
+pub struct WouldBlock;
```
I would lean on the side of no since `TryLockError` in the `poison` module has been around since 1.0 and it has never needed more variants than `Poisoned` and `WouldBlock`?

bors commented:
:umbrella: The latest upstream changes (presumably #144488) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
Okay, this looks great! Thank you for getting this over the line, and @Aandreba for getting it started.

Could you squash "Update UI tests" into "add nonpoison::mutex implementation"? Its description doesn't really make sense on its own. "use generic function instead of macro" should ideally also be squashed somewhere. After that, r=me

connortsui20 committed:
add extra drop, panic, and unwind tests

tgross35 commented:
@bors r-

^ For whatever that is, it's already tested and merged
